### ISSUE TITLE
RSI has different results compared to Talib library

### ISSUE THREAD

**Issue by asimsan** [bukosabino/ta#38]
The RSI calculated from Ta library is different from the one that you get from Trading View. I compared the results between Ta library and Talib library and , Talib gives you the same results as in tradingview.

**Comment by MasterGeorge** [comment]
Same here RSI is not working fine, there are huge diferences

**Comment by rossbonallo** [comment]
Hi, reconciling ta RSI against simple excel compute and Bloomberg it is incorrect.

**Comment by zdr00** [comment]
How different?

Can you provide more details on your excel computation?, remember data might differ from where you get it. Also please take a look at the implementation at `momentum.py`, check if something should be different.

``` 
    diff = close.diff()
    which_dn = diff < 0

    up, dn = diff, diff*0
    up[which_dn], dn[which_dn] = 0, -up[which_dn]

    emaup = ema(up, n, fillna)
    emadn = ema(dn, n, fillna)

    rsi = 100 * emaup / (emaup + emadn)
    if fillna:
        rsi = rsi.replace([np.inf, -np.inf], np.nan).fillna(50)
    return pd.Series(rsi, name='rsi')

```

**Comment by rossbonallo** [comment]
The method used in below sheet rec's with Bloomberg and talib. The error is a country mile. Has not been implemented correctly, nor verified apparently.  

[https://stockcharts.com/school/lib/exe/fetch.php?media=chart_school:technical_indicators_and_overlays:relative_strength_in:cs-rsi.xls](url)

**Comment by fettunch** [comment]
RSI is also not aligned with YAHOO RSI indicator but I think it's a matter on how to tune pandas ewm. 

It all depends on how you specify the decay. Using _com_ rather _span_ and pass n-1 rather than n, it appears to be similar:

    emaup = up.ewm(com=n-1, min_periods=0).mean()
    emadn = dn.ewm(com=n-1, min_periods=0).mean() 

@bukosabino what is your thought on that? what is the golden source for this project to validate the indicators?

**Comment by SebastianB12** [comment]
@fettunch is right. When I'm using:

`emaup = up.ewm(com=n-1, min_periods=0).mean()
emadn = dn.ewm(com=n-1, min_periods=0).mean() `

the RSI values are absolutely identically to Bloomberg (I checked it on my Bloomberg on the S&P 500 Index for the last 20 days), which I would call the golden source, due to the fact, that most traders are using it.
I would very much appreciate a change of the code in this way.

**Comment by juliangall** [comment]
I found this error before finding this thread on github. @fettunch solution doesn't work for me if I compare it with the calculation from [first](https://books.mec.biz/tmp/books/218XOTBWY3FEW2CT3EVR.PDF) principles. It gets gradually more correct as time goes on but is a few percent adrift at the start of the series. I have altered the ema function in ta as follows, based on a Stackoverflow [discussion](https://stackoverflow.com/questions/37924377/does-pandas-calculate-ewm-wrong) about ema in pandas:

`    def ema_alt(series, periods, fillna=False):
        if fillna:
            sma = series.rolling(window=periods, min_periods=0).mean()[:periods]
        else:
            sma = series.rolling(window=periods, min_periods=periods).mean()[:periods]
        rest = series[periods:]
        return pd.concat([sma, rest]).ewm(alpha=1 / periods, adjust=False).mean()
`

**Comment by fettunch** [comment]
thanks @juliangall  for the contribution. As you found there can be different way to compute RSI, that is why I am asking @bukosabino what is the golden source?

**Comment by bukosabino** [comment]
Hi all,

There is not a unique golden source.

The implementation proposed by @fettunch was included in the last version (0.4.7). You can use the new indicator updating the library installation:

```sh
pip install --upgrade ta
```

Thank you all of you.

**Comment by traderblakeq** [comment]
Is there any solution to this ? - Pandas_ta  neither match same value as on Tradingview (TV)
